language/move-vm/runtime/src/loader.rs

```diff
@@ -0,0 +1,1378 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::interpreter_context::InterpreterContext;
+use bytecode_verifier::{
+    verifier::{verify_dependencies, verify_script_dependency_map},
+    VerifiedModule, VerifiedScript,
+};
+use libra_crypto::HashValue;
+use libra_logger::prelude::*;
+use libra_types::{
+    access_path::{AccessPath, Accesses},
+    language_storage::{ModuleId, StructTag, TypeTag},
+    vm_error::{StatusCode, VMStatus},
+};
+use move_core_types::identifier::{IdentStr, Identifier};
+use move_vm_types::loaded_data::{
+    runtime_types::{StructType, Type, TypeConverter},
+    types::{FatStructType, FatType},
+};
+use std::{
+    collections::{BTreeMap, HashMap},
+    fmt::Debug,
+    hash::Hash,
+    sync::{Arc, Mutex},
+};
+use vm::{
+    access::{ModuleAccess, ScriptAccess},
+    errors::{vm_error, Location, VMResult},
+    file_format::{
+        Bytecode, CompiledScript, Constant, ConstantPoolIndex, FieldHandleIndex,
+        FieldInstantiationIndex, FunctionDefinition, FunctionHandleIndex,
+        FunctionInstantiationIndex, Kind, Signature, SignatureToken, StructDefInstantiationIndex,
+        StructDefinition, StructDefinitionIndex, StructFieldInformation,
+    },
+    CompiledModule,
+};
+
+// A simple cache that offers both a HashMap and a Vector lookup.
+// Values are forced into a `Arc` so they can be used from multiple thread.
+// Access to this cache is always under a `Mutex`.
+struct BinaryCache<K, V> {
+    id_map: HashMap<K, usize>,
+    binaries: Vec<Arc<V>>,
+}
+
+impl<K, V> BinaryCache<K, V>
+where
+    K: Eq + Hash,
+{
+    fn new() -> Self {
+        Self {
+            id_map: HashMap::new(),
+            binaries: vec![],
+        }
+    }
+
+    fn insert(&mut self, key: K, binary: V) -> &Arc<V> {
+        self.binaries.push(Arc::new(binary));
+        let idx = self.binaries.len() - 1;
+        self.id_map.insert(key, idx);
+        self.binaries
+            .last()
+            .expect("BinaryCache: last() after push() impossible failure")
+    }
+
+    fn get(&self, key: &K) -> Option<&Arc<V>> {
+        self.id_map
+            .get(&key)
+            .and_then(|idx| self.binaries.get(*idx))
+    }
+}
```

Comment:
fair point and something to review in a near future, I am not clear yet if we need an index one and also if this may be used for types and functions which right now are searched with a loop over the vector.
So not sure where this will land...